pkg/flow/internal/controller/loader.go

```diff
@@ -168,31 +170,49 @@ func (l *Loader) Apply(args map[string]any, componentBlocks []*ast.BlockStmt, co
 
 		var err error
 
-		switch c := n.(type) {
+		switch n := n.(type) {
 		case *ComponentNode:
-			components = append(components, c)
-			componentIDs = append(componentIDs, c.ID())
+			components = append(components, n)
+			componentIDs = append(componentIDs, n.ID())
 
-			if err = l.evaluate(logger, c); err != nil {
+			if err = l.evaluate(logger, n); err != nil {
 				var evalDiags diag.Diagnostics
 				if errors.As(err, &evalDiags) {
 					diags = append(diags, evalDiags...)
 				} else {
 					diags.Add(diag.Diagnostic{
 						Severity: diag.SeverityLevelError,
 						Message:  fmt.Sprintf("Failed to build component: %s", err),
-						StartPos: ast.StartPos(c.Block()).Position(),
-						EndPos:   ast.EndPos(c.Block()).Position(),
+						StartPos: ast.StartPos(n.Block()).Position(),
+						EndPos:   ast.EndPos(n.Block()).Position(),
 					})
 				}
 			}
+
+		case *ServiceNode:
+			services = append(services, n)
+
+			if err = l.evaluate(logger, n); err != nil {
```

Comment:
I've been pondering on this for a few weeks now. I think in the future we might want to try to adjust the order of the walk to put certain critical nodes first (i.e., services, logging/tracing), but for now it doesn't really matter since the services will always evaluate before any node which depends on them.